sys/kern/vfs_bio.c

```diff
@@ -1140,9 +1142,15 @@ kern_vfs_bio_buffer_alloc(caddr_t v, long physmem_est)
 
 	/*
 	 * Reserve space for the buffer cache buffers
+	 * When we are called the first time, the capability is invalid
+	 * so we can not set bounds.
 	 */
-	buf = (void *)v;
-	v = (caddr_t)(buf + nbuf);
+	if (!cheri_kern_gettag(v)) {
+		return (v + nbuf * sizeof(*buf));
+	}
+
+	buf = (void *)cheri_kern_setbounds(v, nbuf * sizeof(*buf));
+	v = v + nbuf * sizeof(*buf);
```

Comment:
```suggestion
	if (cheri_kern_gettag(v))
		buf = (void *)cheri_kern_setbounds(v, nbuf * sizeof(*buf));

	v = v + nbuf * sizeof(*buf);
```
though this code is rather weird and it feels like this is masking a problem elsewhere?

Reply:
This function is horrible, it gets called two times: the first time with "invalid" arguments so that it calculates the allocation sizes and the second one actually performs the allocation. This is happening before of the two pass approach that is being taken in vm_init `vm_ksubmap_init()`. The whole thing should be reworked upstream but this change keeps our diff minimal here.